metrics/src/metrics_manager.rs

```diff
@@ -0,0 +1,135 @@
+        }).expect("Failed to spawn websocket thread");
+        info!("Starting websocket server at port: {}", socket_port);
+
+
+        Self {
+            _socket_port: socket_port,
+            event_channel,
+            broadcaster,
+            connected_clients,
+            peer_monitors: HashMap::new(),
+        }
+    }
+
+    pub fn actor(sys: &impl ActorRefFactory, event_channel: ChannelRef<NetworkChannelMsg>, socket_port: u16) -> Result<MetricsManagerRef, CreateError> {
+        sys.actor_of(
+            Props::new_args(Self::new, (event_channel, socket_port)),
+            Self::name(),
+        )
+    }
+}
+
+impl Actor for MetricsManager {
+    type Msg = MetricsManagerMsg;
+
+    fn pre_start(&mut self, ctx: &Context<Self::Msg>) {
+        // Listen for all network events
+        self.event_channel.tell(Subscribe {
+            actor: Box::new(ctx.myself()),
+            topic: NetworkChannelTopic::NetworkEvents.into(),
+        }, None);
+
+        // Every second, send yourself a message to broadcast the metrics to all connected clients
+        ctx.schedule(Duration::from_secs(1),
+                     Duration::from_secs(1),
+                     ctx.myself(), None,
+                     BroadcastSignal);
+    }
+
+    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, sender: Option<BasicActorRef>) {
+        self.receive(ctx, msg, sender);
+    }
+}
+
+impl Receive<BroadcastSignal> for MetricsManager {
+    type Msg = MetricsManagerMsg;
+
+    fn receive(&mut self, _ctx: &Context<Self::Msg>, _msg: BroadcastSignal, _sender: Sender) {
+        use serde_json::to_string;
+        use std::sync::atomic::Ordering::Relaxed;
+
+        if self.connected_clients.load(Relaxed) > 0 {
+            // There are clients connected to the WebSocket
+            let snapshot_data: HashMap<&String, PeerMetrics> = self.peer_monitors.iter_mut()
+                .map(|(k, v)| {
+                    (k, v.snapshot())
+                }).collect();
+            if let Ok(message) = to_string(&snapshot_data) {
+                if let Err(err) = self.broadcaster.send(message) {
+                    warn!("Failed broadcast message: {}", err);
+                }
+            } else {
+                warn!("Failed to serialize snapshot data");
+            }
+        }
+    }
+}
+
+impl Receive<NetworkChannelMsg> for MetricsManager {
+    type Msg = MetricsManagerMsg;
+
+    fn receive(&mut self, _ctx: &Context<Self::Msg>, msg: NetworkChannelMsg, _sender: Sender) {
+        use std::mem::size_of_val;
+        match msg {
+            NetworkChannelMsg::PeerCreated(msg) => {
+                // TODO: Add field to message with peer name / identifier
+                let peer_name = msg.peer.name().to_owned();
+                let monitor = PeerMonitor::new(peer_name.clone());
+                self.peer_monitors.insert(peer_name, monitor);
+            }
+            NetworkChannelMsg::PeerDisconnected(msg) => {
```

Comment:
Instead of peer_disconnected you can listen to system event like we do here https://github.com/simplestaking/tezos-rs/blob/master/shell/src/peer_manager.rs#L103-L107

Reply:
Fixed